### PR TITLE
ST-299 ST-333 - Irregular hours journey with updated text

### PR DESCRIPTION
**Pull request by amaalali**
- [x] Unit tests passing
- [x] Coverage reached
- [x] Acceptance tests passing
- [x] Reviewed

### CODE REVIEW

File: app/controllers/income/estimatedPay/update/IncomeUpdateCalculatorController.scala [hunk 1]
```diff
@@ -31,24 +32,28 @@ import uk.gov.hmrc.renderer.TemplateRenderer
...
 import uk.gov.hmrc.tai.service._
+import uk.gov.hmrc.tai.util.constants.EditIncomeIrregularPayConstants
 import uk.gov.hmrc.tai.util.{FormHelper, JourneyCacheConstants}
+import uk.gov.hmrc.tai.viewModels.income.{ConfirmIncomeIrregularHoursViewModel, EditIncomeIrregularHoursViewModel}
 import uk.gov.hmrc.tai.viewModels.income.estimatedPay.update.CheckYourAnswersViewModel
 
 import scala.Function.tupled
 import scala.concurrent.Future
 
+//noinspection ScalaStyle
```
Comment (SeanArmstrong) [inline app/controllers/income/estimatedPay/update/IncomeUpdateCalculatorController.scala]: To remove

File: app/controllers/income/estimatedPay/update/IncomeUpdateCalculatorController.scala [hunk 2]
```diff
@@ -170,15 +175,97 @@ trait IncomeUpdateCalculatorController extends TaiBaseController
...
-            formData.workingHours match {
-              case Some("same") => Future.successful(Redirect(routes.IncomeUpdateCalculatorController.payPeriodPage()))
-              case _ => Future.successful(Redirect(routes.IncomeUpdateCalculatorController.calcUnavailablePage()))
+          (formData: HoursWorkedForm) => {
+            val idRequest = journeyCacheService.mandatoryValueAsInt(UpdateIncome_IdKey)
+            val employerNameRequest = journeyCacheService.mandatoryValue(UpdateIncome_NameKey)
+
+            for {
+              id <- idRequest
+              employerName <- employerNameRequest
```
Comment (SeanArmstrong) [inline app/controllers/income/estimatedPay/update/IncomeUpdateCalculatorController.scala]: This is not used

File: app/views/incomes/confirmIncomeIrregularHours.scala.html [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+@import includes.forms.submit
+@import uk.gov.hmrc.tai.viewModels.income.ConfirmIncomeIrregularHoursViewModel
+
+@(viewModel: ConfirmIncomeIrregularHoursViewModel)(
+    implicit request: Request[_],
+    messages: Messages,
+    templateRenderer: uk.gov.hmrc.renderer.TemplateRenderer,
+    partialRetriever: uk.gov.hmrc.play.partials.FormPartialRetriever)
+
+@gaEventActionString = @{"???"}
```
Comment (SeanArmstrong) [inline app/views/incomes/confirmIncomeIrregularHours.scala.html]: This need populating?

File: app/views/incomes/editIncomeIrregularHours.scala.html [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+@import includes.forms.submit
+
+@(incomeUpdateForm: Form[uk.gov.hmrc.tai.forms.EditIncomeIrregularHoursForm],
+  viewModel: EditIncomeIrregularHoursViewModel)(
+    implicit request: Request[_],
+    messages: Messages,
+    templateRenderer: uk.gov.hmrc.renderer.TemplateRenderer,
+    partialRetriever: uk.gov.hmrc.play.partials.FormPartialRetriever)
+
+@gaEventActionString = @{"???"}
```
Comment (SeanArmstrong) [inline app/views/incomes/editIncomeIrregularHours.scala.html]: This need populating?

File: app/views/incomes/confirmIncomeIrregularHours.scala.html [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+    <div class="panel panel-border-narrow">
+        <p>@messages("tai.irregular.confirm.effectOfChange")</p>
+    </div>
+
+
+
+    <p><a id="confirmAndSend" class="button section" role="button" href="@viewModel.submitUrl">@messages("tai.confirmAndSend")</a></p>
+
+    <p><a href="@routes.IncomeSourceSummaryController.onPageLoad(2)">@messages("tai.cancel.noSave")</a></p>
```
Comment (SeanArmstrong) [inline app/views/incomes/confirmIncomeIrregularHours.scala.html]: Strange route? What does '2' signify?

File: test/controllers/income/estimatedPay/update/IncomeUpdateCalculatorControllerSpec.scala [hunk 6]
```diff
@@ -612,16 +629,305 @@ class IncomeUpdateCalculatorControllerSpec extends PlaySpec with FakeTaiPlayAppl
...
+      when(
+        testController.journeyCacheService.cache(any(), any())(any())
+      ).thenReturn(
+        Future.successful(Map.empty[String, String])
+      )
+
+      val result = testController.handleIncomeIrregularHours(1)(FakeRequest(method = "POST", path = "")
+        .withFormUrlEncodedBody("income" -> "999")
+        .withSession(
+          SessionKeys.sessionId -> s"session-${UUID.randomUUID()}",
```
Comment (SeanArmstrong) [inline test/controllers/income/estimatedPay/update/IncomeUpdateCalculatorControllerSpec.scala]: This session id map could be extracted to somewhere else? Is this common in alot of FakeRequests?

File: app/views/incomes/confirmIncomeIrregularHours.scala.html (hunk 5 above)
Comment (amaalali) [reply app/views/incomes/confirmIncomeIrregularHours.scala.html]: Good spot this was supposed to be the employer id

File: test/controllers/income/estimatedPay/update/IncomeUpdateCalculatorControllerSpec.scala (hunk 6 above)
Comment (amaalali) [reply test/controllers/income/estimatedPay/update/IncomeUpdateCalculatorControllerSpec.scala]: 👍 I've added a builder to `RequestBuilder` for this

File: test/builders/RequestBuilder.scala [hunk 7]
```diff
@@ -18,11 +18,24 @@ package builders
...
 
+import play.api.mvc.AnyContentAsFormUrlEncoded
 import play.api.test.FakeRequest
 import uk.gov.hmrc.http.SessionKeys
 
 object RequestBuilder {
 
+  private val HTTP_VERBS = List("GET", "POST", "PUT", "DELETE")
+
+  def buildFaleReqestWithOnlySession(method: String) = {
```
Comment (SeanArmstrong) [inline test/builders/RequestBuilder.scala]:
```suggestion
  def buildFakeReqestWithOnlySession(method: String) = {
```

File: app/uk/gov/hmrc/tai/viewModels/income/EditIncomeIrregularHoursViewModel.scala [hunk 8]
```diff
@@ -17,18 +17,31 @@
...
             employerName: String,
-            currentAmount: Int)(implicit messages: Messages): EditIncomeIrregularHoursViewModel = {
+            currentAmount: String)(implicit messages: Messages): EditIncomeIrregularHoursViewModel = {
+
+    val taxYearMessage = currentTaxYearRangeHtmlNonBreak
+
+    val heading = messages("tai.irregular.mainHeadingText", taxYearMessage)
+    val preHeading = messages("tai.estimatedPay.preHeading", employerName)
+
+    val currentAmountBigDecimal: String = withPoundPrefix(MoneyPounds(BigDecimal(currentAmount),0))
```
Comment (SeanArmstrong) [inline app/uk/gov/hmrc/tai/viewModels/income/EditIncomeIrregularHoursViewModel.scala]: 👍 Should be using this more

**Pull request merged**